### PR TITLE
Loyaltyprogram

### PR DESCRIPTION
**Pull request by gd-kkramarenko**

### CODE REVIEW

File: EcommerceApp/backend/spring-boot-ecommerce-app/src/main/java/kkramarenko/ecommerceapp/enums/DiscountRate.java [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package kkramarenko.ecommerceapp.enums;
+
+public enum DiscountRate {
+    starter,
+    bronze,
+    silver,
+    gold,
+    platinum;
```
Comment (gd-iistomin) [inline EcommerceApp/backend/spring-boot-ecommerce-app/src/main/java/kkramarenko/ecommerceapp/enums/DiscountRate.java:8]: enums should be written in capital letters

File: EcommerceApp/backend/spring-boot-ecommerce-app/src/main/java/kkramarenko/ecommerceapp/service/CheckoutServiceImpl.java [hunk 2]
```diff
@@ -62,17 +75,35 @@ public PurchaseResponse placeOrder(Purchase purchase) {
         Customer existingCustomer = customerRepository.findCustomerByFirstNameAndLastNameAndEmail(
                     customer.getFirstName(), customer.getLastName(), customer.getEmail());
 
+        boolean discountRateChanged = false;
+        String newDiscountRate = "";
+
         if(existingCustomer != null){
```
Comment (gd-iistomin) [inline EcommerceApp/backend/spring-boot-ecommerce-app/src/main/java/kkramarenko/ecommerceapp/service/CheckoutServiceImpl.java:81]: Code formatting:
`if (existingCustomer != null) {`

Comment (gd-iistomin) [reply EcommerceApp/backend/spring-boot-ecommerce-app/src/main/java/kkramarenko/ecommerceapp/service/CheckoutServiceImpl.java:81]: You also can try to autoformat your code by using this feature in IDE: https://www.jetbrains.com/help/idea/reformat-and-rearrange-code.html#reformat_file

File: EcommerceApp/backend/spring-boot-ecommerce-app/src/main/java/kkramarenko/ecommerceapp/service/CheckoutServiceImpl.java [hunk 3]
```diff
@@ -62,17 +75,35 @@ public PurchaseResponse placeOrder(Purchase purchase) {
...
+            User userWithGivenCustomer = userRepository.findUserByCustomer(existingCustomer);
+            if(userWithGivenCustomer == null) {
+                existingCustomer.add(order);
+                customerRepository.save(existingCustomer);
+            } else {
+                DiscountRate discountRateBeforePurchase = DiscountRate.valueOf(userWithGivenCustomer.getDiscountRate());
+                existingCustomer.add(order);
+                customerRepository.save(existingCustomer);
+                DiscountRate discountRateAfterPurchase = loyaltyProgramService.getCustomerCurrentDiscountRate(existingCustomer);
+                if(!(discountRateAfterPurchase.equals(discountRateBeforePurchase))) {
```
Comment (gd-iistomin) [inline EcommerceApp/backend/spring-boot-ecommerce-app/src/main/java/kkramarenko/ecommerceapp/service/CheckoutServiceImpl.java:91]: condition can be replaced with
`(!discountRateAfterPurchase.equals(discountRateBeforePurchase))`

File: EcommerceApp/backend/spring-boot-ecommerce-app/src/main/java/kkramarenko/ecommerceapp/service/LoyaltyProgramServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        List<Order> customerOrders = orderRepository.findOrdersByCustomer(customer);
+
+        if(customerOrders.size() == 0) { return DiscountRate.starter; }
+
+        BigDecimal customerOrdersTotalPriceSum = new BigDecimal("0.00");
+        for(Order order: customerOrders){
+            customerOrdersTotalPriceSum = customerOrdersTotalPriceSum.add(order.getTotalPrice());
+        }
+
+        if(customerOrdersTotalPriceSum.compareTo(new BigDecimal("100.00")) < 0){
```
Comment (gd-iistomin) [inline EcommerceApp/backend/spring-boot-ecommerce-app/src/main/java/kkramarenko/ecommerceapp/service/LoyaltyProgramServiceImpl.java:43]: The best place for such constants is database, not the code. Actually it takes effort to implement so there are other options:
for example, you can declare these numbers as named constants or as fields in DiscountRate enum

**Review by gd-iistomin — CHANGES_REQUESTED** [review]

File: EcommerceApp/backend/spring-boot-ecommerce-app/src/main/java/kkramarenko/ecommerceapp/enums/DiscountRate.java (hunk 1 above)
Comment (gd-kkramarenko) [reply EcommerceApp/backend/spring-boot-ecommerce-app/src/main/java/kkramarenko/ecommerceapp/enums/DiscountRate.java:8]: Done

File: EcommerceApp/backend/spring-boot-ecommerce-app/src/main/java/kkramarenko/ecommerceapp/service/CheckoutServiceImpl.java (hunk 2 above)
Comment (gd-kkramarenko) [reply EcommerceApp/backend/spring-boot-ecommerce-app/src/main/java/kkramarenko/ecommerceapp/service/CheckoutServiceImpl.java:81]: Done

File: EcommerceApp/backend/spring-boot-ecommerce-app/src/main/java/kkramarenko/ecommerceapp/service/CheckoutServiceImpl.java (hunk 3 above)
Comment (gd-kkramarenko) [reply EcommerceApp/backend/spring-boot-ecommerce-app/src/main/java/kkramarenko/ecommerceapp/service/CheckoutServiceImpl.java:91]: Done

File: EcommerceApp/backend/spring-boot-ecommerce-app/src/main/java/kkramarenko/ecommerceapp/service/LoyaltyProgramServiceImpl.java (hunk 4 above)
Comment (gd-kkramarenko) [reply EcommerceApp/backend/spring-boot-ecommerce-app/src/main/java/kkramarenko/ecommerceapp/service/LoyaltyProgramServiceImpl.java:43]: Done

**Review by gd-iistomin — APPROVED** [review]

**Pull request merged**